### ISSUE TITLE
[FR] Model registry semantic versioning

### ISSUE THREAD

**Issue by addisonklinke** [mlflow/mlflow#7280]
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

For registered model versions allow user-specific semantic versioning strings (i.e. `v1.1.3`) in addition to auto-incrementing integers

### Motivation

> #### What is the use case for this feature?

Providing greater flexibility to users for identifying a registered model

> #### Why is this use case valuable to support for MLflow users in general?

Currently, it's not possible to indicate *how much* a model has changed between versions because the version number increment is always the same. Semantic versioning would provide a much richer description of each model that makes it easier for engineers to recall what's what without having to click through to the source run page. Also, semantic versions could integrate better with external projects (i.e. matching the git tag / release version of a Github repository, etc)

> #### Why is this use case valuable to support for your project(s) or organization?

Our team could create conventions for the different components of the version number. For example
* **Major:** new architecture
* **Minor:** newly trained weights
* **Patch:** inference format change (i.e. fused conv/batchnorm, using a new framework, etc) while weights/architecture are constant
    * This happens frequently when deploying to edge devices because we may find small tweaks that help layers run more efficiently or enable access to accelerator chips

> #### Why is it currently difficult to achieve this use case?

There's no option to specify any version number other than the default, auto-incrementing integers 

### Details

How might this look in the UI?
1. Click "Register Model" button from the run page artifacts section
2. Select the model name
3. A default version number is populated, but the user can change it 

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [X] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [X] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by BenWilson2** [comment]
Could the same effect be realized with using tags to define the differences in change magnitude?

**Comment by addisonklinke** [comment]
I think that's a potential workaround with a few downsides
* Version number is an argument to several Python client methods including `MlfowClient.update_model_version()`. This would be critical for integrating with external services
* Many parts of the UI would not respond to the tagged semantic version (i.e. registered model page)
* Tags can be edited/deleted independent of the registered model object whereas the true model version cannot be changed after creation (more resistant to accidental tampering)
* It feels odd to use a tag key for a model property which already exists. This might cause confusion in team communication

I understand this might be lower priority, but it still seems like a useful feature in the long-run

**Comment by cristofp** [comment]
If I can add some suggestion to this FR:

In my company we are designing automatic training and deployment of fresh models in production.
In this case we need to have versioning pattern **that will let for disthinguishing following changes in the model**:
- a) new version of model breaks backward compatibility (inference method new signature is not backward compatible)
- b) new version of model is backward compatible, but there was some code change in ML Project
- c) new version of model is produced using the same ML Project code, but with differed (newer) training data set

This distinction is fully coherent with Semantic Versioning. And accordingly using `major.minor.patch` pattern we will:
 - a) bump `major` number when breaking backward compatibility
 - b) bump `minor` number when preserving backward compatibility but introducing code changes in ML Project
 - c) bump `patch` number when just retraining model using the same ML Project code

**Comment by addisonklinke** [comment]
@cristofp yes exactly! I think many production workflows need more granularity in the version number beyond a single integer. This is standard practice in traditional software, so it makes a lot of sense to provide the same capability for ML models

**Comment by uduwage** [comment]
We have created a work around solution without really changing the actual version. We created a solution with the use of bump version and using semantic versioning. We would love to make PR to Mlflow to expose that feature. I do believe ultimately we want to be able to change the "actual" version `(version 1 --> 1.0.0 etc)` but we needed this right and happy to create a PR. Any thoughts? guidance?

**Comment by AsmaDagmoune** [comment]
Hello, Do you have any updated regarding this particular point please, I see that the ticket is still open :(
Thanks in advance ^^
